From 497f1da1d4fe5e05ed2a4dde9e87ecb5a5c24a04 Mon Sep 17 00:00:00 2001
From: Bo Yang <bo.yang@amlogic.com>
Date: Tue, 26 May 2015 17:17:22 +0800
Subject: [PATCH 5617/5965] PD#106651: amports: sync vdec clk of g9bb with
 m8baby

g9bb hevc clk is same with m8bay that reviewed by boren.deng.

Change-Id: I62b35b338f40bb53a753322fa5e35f4bf97634fd
---
 drivers/amlogic/amports/Kconfig         |  2 +-
 drivers/amlogic/amports/g9bb/vdec_clk.c | 55 ++++++++++---------------
 drivers/amlogic/amports/vh265.c         |  2 +
 3 files changed, 24 insertions(+), 35 deletions(-)

diff --git a/drivers/amlogic/amports/Kconfig b/drivers/amlogic/amports/Kconfig
index e678fb6a834e..368c07b614da 100755
--- a/drivers/amlogic/amports/Kconfig
+++ b/drivers/amlogic/amports/Kconfig
@@ -112,7 +112,7 @@ config AM_VDEC_H264_4K2K
 
 config AM_VDEC_H265
         tristate "Amlogic H265 Video Decoder"
-        depends on ARCH_MESON8B || ARCH_MESON8 || ARCH_MESONG9TV
+        depends on ARCH_MESON8B || ARCH_MESON8 || ARCH_MESONG9TV || ARCH_MESONG9BB
         default y
 
         select AM_PTSSERVER
diff --git a/drivers/amlogic/amports/g9bb/vdec_clk.c b/drivers/amlogic/amports/g9bb/vdec_clk.c
index 67af3fa6b5d5..3fd389dc1464 100644
--- a/drivers/amlogic/amports/g9bb/vdec_clk.c
+++ b/drivers/amlogic/amports/g9bb/vdec_clk.c
@@ -71,33 +71,17 @@ HHI_VDEC_CLK_CNTL
 //364.29M <-- (2550/7)/1 -- over limit, do not use
 #define VDEC1_364M() WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  (3 << 9) | (0), 0, 16)
 #define VDEC2_364M() WRITE_MPEG_REG(HHI_VDEC2_CLK_CNTL, (3 << 9) | (0))
+#define HEVC_364M()  WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, (3 << 9) | (0), 16, 16)
 
-//425.00M <-- (2550/3)/2
-#define VDEC1_425M() WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  (1 << 9) | (2), 0, 16)
-
-//510.00M <-- (2550/5)/1
-#define VDEC1_510M() WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  (2 << 9) | (0), 0, 16)
-#define HEVC_510M()  WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, (2 << 9) | (0), 16, 16)
-
-//637.50M <-- (2550/4)/1
-#define VDEC1_638M() WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  (0 << 9) | (0), 0, 16)
-#define HEVC_638M()  WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, (0 << 9) | (0), 16, 16)
-
-#define VDEC1_CLOCK_ON()  do { if (IS_MESON_M8_CPU) { \
-                                    WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL, 1, 8, 1); \
-                                    WRITE_VREG_BITS(DOS_GCLK_EN0, 0x3ff,0,10); \
-                                } else { \
-                                    WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL, 1, 8, 1); \
-                                    WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 15, 1); \
-                                    WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 8, 1); \
-                                    WRITE_VREG_BITS(DOS_GCLK_EN0, 0x3ff,0,10); \
-                                } \
-                            } while (0)
-
+#define VDEC1_CLOCK_ON()   WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL, 1, 8, 1); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 15, 1); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 8, 1); \
+                           WRITE_VREG_BITS(DOS_GCLK_EN0, 0x3ff,0,10)
 #define VDEC2_CLOCK_ON()   WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, 1, 8, 1); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 31, 1); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 0, 24, 1); \
                            WRITE_VREG(DOS_GCLK_EN1, 0x3ff)
-
-#define HCODEC_CLOCK_ON()  WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL, 1, 24, 1); \
+#define HCODEC_CLOCK_ON()  WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  1, 24, 1); \
                            WRITE_VREG_BITS(DOS_GCLK_EN0, 0x7fff, 12, 15)
 #define HEVC_CLOCK_ON()    WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, 0, 24, 1); \
                            WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, 0, 31, 1); \
@@ -106,6 +90,9 @@ HHI_VDEC_CLK_CNTL
 #define VDEC1_SAFE_CLOCK() WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, READ_MPEG_REG(HHI_VDEC_CLK_CNTL) & 0x7f, 0, 7); \
                            WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 1, 8, 1); \
                            WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 1, 15, 1);
+#define VDEC2_SAFE_CLOCK() WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, (READ_MPEG_REG(HHI_VDEC_CLK_CNTL) >> 16) & 0x7f, 16, 7); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 1, 24, 1); \
+                           WRITE_MPEG_REG_BITS(HHI_VDEC3_CLK_CNTL, 1, 31, 1);
 #define VDEC1_CLOCK_OFF()  WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL,  0, 8, 1)
 #define VDEC2_CLOCK_OFF()  WRITE_MPEG_REG_BITS(HHI_VDEC2_CLK_CNTL, 0, 8, 1)
 #define HCODEC_CLOCK_OFF() WRITE_MPEG_REG_BITS(HHI_VDEC_CLK_CNTL, 0, 24, 1)
@@ -119,7 +106,7 @@ static int clock_level[VDEC_MAX+1];
 void vdec_clock_enable(void)
 {
     VDEC1_CLOCK_OFF();
-    VDEC1_255M();
+    VDEC1_182M();
     VDEC1_CLOCK_ON();
     clock_level[VDEC_1] = 0;
 }
@@ -127,11 +114,7 @@ void vdec_clock_enable(void)
 void vdec_clock_hi_enable(void)
 {
     VDEC1_CLOCK_OFF();
-    if (IS_MESON_M8_CPU) {
-        VDEC1_364M();
-    } else {
-        VDEC1_638M();
-    }
+    VDEC1_364M();
     VDEC1_CLOCK_ON();
     clock_level[VDEC_1] = 1;
 }
@@ -149,7 +132,7 @@ void vdec_clock_off(void)
 void vdec2_clock_enable(void)
 {
     VDEC2_CLOCK_OFF();
-    VDEC2_255M();
+    VDEC2_182M();
     VDEC2_CLOCK_ON();
     clock_level[VDEC_2] = 0;
 }
@@ -192,15 +175,14 @@ void hcodec_clock_off(void)
 void hevc_clock_enable(void)
 {
     HEVC_CLOCK_OFF();
-//    HEVC_255M();
-    HEVC_638M();
+    HEVC_255M();
     HEVC_CLOCK_ON();
 }
 
 void hevc_clock_hi_enable(void)
 {
     HEVC_CLOCK_OFF();
-    HEVC_638M();
+    HEVC_364M();
     HEVC_CLOCK_ON();
     clock_level[VDEC_HEVC] = 1;
 }
@@ -220,6 +202,11 @@ void vdec_clock_prepare_switch(void)
     VDEC1_SAFE_CLOCK();
 }
 
+void vdec2_clock_prepare_switch(void)
+{
+    VDEC2_SAFE_CLOCK();
+}
+
 void hevc_clock_prepare_switch(void)
 {
     HEVC_SAFE_CLOCK();
diff --git a/drivers/amlogic/amports/vh265.c b/drivers/amlogic/amports/vh265.c
index d525e6b0a967..586e6070c8d2 100755
--- a/drivers/amlogic/amports/vh265.c
+++ b/drivers/amlogic/amports/vh265.c
@@ -51,6 +51,8 @@
 #define USE_BUF_BLOCK
 #if MESON_CPU_TYPE == MESON_CPU_TYPE_MESON8B
 #undef SUPPORT_4K2K
+#elif MESON_CPU_TYPE == MESON_CPU_TYPE_MESONG9BB
+#undef SUPPORT_4K2K
 #else
 #define SUPPORT_4K2K
 #endif
-- 
2.19.0

